cayenne-server/src/main/java/org/apache/cayenne/access/ObjectStore.java

```diff
@@ -213,13 +217,28 @@ synchronized ObjectDiff registerDiff(Object nodeId, NodeDiff diff) {
         return objectDiff;
     }
 
+    private Persistent getUnwrapped(Object nodeId) {
+        Persistent persistent = objectMap.get(nodeId);
+        if(persistent == null) {
+            return null;
+        }
+        return ((ObjectStorePersistentWrapper) persistent).dataObject();
+    }
+
     /**
      * Returns a number of objects currently registered with this ObjectStore.
      * 
      * @since 1.2
      */
     public int registeredObjectsCount() {
-        return objectMap.size();
+        AtomicInteger counter = new AtomicInteger();
+        objectMap.forEach((id, obj) -> {
+            ObjectStorePersistentWrapper wrapper = (ObjectStorePersistentWrapper) obj;
+            if(wrapper.hasObject()){
+                counter.incrementAndGet();
+            }
+        });
+        return counter.get();
```

Comment:
Alternative instead of using AtomicInteger maybe ?
```java
return (int) objectMap.values().stream()
        .map( obj ->  (ObjectStorePersistentWrapper) obj )
        .filter( wrap -> wrap.hasObject() )
        .count();
```

Reply:
In any other place I would go with streams too in a such method. But in the Cayenne core we are trying to keep performance in mind, and leave streams out.

Reply:
Okay :-)  Speaking of performance though I'm concerned with all the casting happening in this PR, or am I mistaken ?

Reply:
Static cast is a compile-time feature, so it should be ok. But additional wrapper would add some overhead unfortunately.